Source/DafnyLanguageServer.Test/Lookup/DefinitionTest.cs

```diff
@@ -29,200 +29,209 @@ public class DefinitionTest : ClientBasedLanguageServerTest {
+      Assert.AreEqual(ranges[1], eReference.Location!.Range);
 
-      var fooReference = (await RequestDefinition(documentItem, (7, 8)).AsTask()).Single();
-      Assert.AreEqual(new Range((2, 27), (2, 30)), fooReference.Location!.Range);
+      var fooReference = (await RequestDefinition(documentItem, positions[2]).AsTask()).Single();
+      Assert.AreEqual(ranges[2], fooReference.Location!.Range);
     }
 
     [TestMethod]
     public async Task FunctionCallAndGotoOnDeclaration() {
       var source = @"
-function FibonacciSpec(n: nat): nat {
+function [|Fibo$$nacciSpec|]($$n: nat): nat {
   if (n == 0) then 0
   else if (n == 1) then 1
-  else FibonacciSpec(n - 1) + FibonacciSpec(n - 2)
+  else Fi$$bonacciSpec(n - 1) + FibonacciSpec(n - 2)
 }
 
-type seq31<T> = x: seq<T> | 0 <= |x| <= 32 as int
+type seq31<[|T|]> = x: seq<$$T> | 0 <= |x| <= 32 as int
 ".TrimStart();
 
-      var documentItem = CreateTestDocument(source);
+      MarkupTestFile.GetPositionsAndRanges(source, out var cleanSource,
+        out var positions, out var ranges);
+      var documentItem = CreateTestDocument(cleanSource);
       await client.OpenDocumentAndWaitAsync(documentItem, CancellationToken);
-      var definition = (await RequestDefinition(documentItem, (3, 8)).AsTask()).Single();
-      var location = definition.Location;
-      Assert.AreEqual(documentItem.Uri, location.Uri);
-      Assert.AreEqual(new Range((0, 9), (0, 22)), location.Range);
 
-      var fibonacciSpecOnItself = (await RequestDefinition(documentItem, (0, 12)).AsTask());
+      var fibonacciSpecOnItself = (await RequestDefinition(documentItem, positions[0]).AsTask());
       Assert.IsFalse(fibonacciSpecOnItself.Any());
 
-      var nOnItself = (await RequestDefinition(documentItem, (0, 23)).AsTask());
+      var nOnItself = (await RequestDefinition(documentItem, positions[1]).AsTask());
       Assert.IsFalse(nOnItself.Any());
 
-      var typeParameter = (await RequestDefinition(documentItem, (6, 23)).AsTask()).Single();
-      Assert.AreEqual(new Range((6, 11), (6, 12)), typeParameter.Location!.Range);
+      var fibonacciCall = (await RequestDefinition(documentItem, positions[2]).AsTask()).Single();
+      Assert.AreEqual(ranges[0], fibonacciCall.Location!.Range);
+
+      var typeParameter = (await RequestDefinition(documentItem, positions[3]).AsTask()).Single();
+      Assert.AreEqual(ranges[1], typeParameter.Location!.Range);
     }
 
     [TestMethod]
     public async Task DatatypesAndMatches() {
       var source = @"
-datatype Identity<T> = Identity(value: T)
-datatype Colors = Red | Green | Blue
+datatype Identity<T> = [|Identity|](value: T)
+datatype Colors = Red | [|Green|] | Blue
 
-function Foo(value: Identity<Colors>): bool {
-  match value {
-    case Identity(Red()) => true
-    case Identity(Green) => false // Warning
+function Foo([|value|]: Identity<Colors>): bool {
+  match va$$lue {
+    case Ide$$ntity(Red()) => true
+    case Identity(Gr$$een) => false // Warning
```

Comment:
I love this new test format ! I can finally understand what each position and range is about :-)
I also see that you can extract named spans in your code, but I don't see them yet in tests....
So I would like to challenge you on this second but final improvement: Rather on referring to these ranges and positions by `ranges[i]` and `positions[j]`, could you use named ranges instead?

```suggestion
function Foo({|0:value|}: Identity<Colors>): bool {
  match va$0$lue {
    case Ide$$ntity(Red()) => true
    case Identity(Gr$$een) => false // Warning
```

The point of using named ranges (and thus named positions), is that your test becomes self-contained in the source: You can just iterate over named positions, query for the definition range, and verify it corresponds to the same named range.
It makes adding new tests extremely easy.

Reply:
Sounds good :)